lib/driver.js

```diff
@@ -244,25 +279,52 @@ class EspressoDriver extends BaseDriver {
       }
     };
 
-    const isApk = _.endsWith(_.toLower(appPath), '.apk');
-    // Do not cache .apk if was already present on the local file system
-    const shouldResultAppPathBeCached = !isApk || (isApk && isUrl);
+    const hasApkExt = (appPath) => _.endsWith(_.toLower(appPath), APK_EXT);
+    const extractUniversalApk = async (shouldExtract, appPath) =>
+      shouldExtract ? appPath : await this.adb.extractUniversalApk(appPath);
+
     let pathInCache = null;
     let isResultAppPathAlreadyCached = false;
     if (_.isPlainObject(cachedAppInfo)) {
       const packageHash = await fs.hash(appPath);
       if (packageHash === cachedAppInfo.packageHash && await fs.exists(cachedAppInfo.fullPath)) {
+        logger.info(`Using '${cachedAppInfo.fullPath}' that is cached from '${appPath}'`);
         isResultAppPathAlreadyCached = true;
         pathInCache = cachedAppInfo.fullPath;
       }
     }
+
+    // appPath can be .zip, .apk, .apks or .aab
+    const isApk = hasApkExt(appPath);
+    // local .zip, .aab and downloaded .zip, .aab and .apks should be cached
+    // to skip unzip and generate .apk from .aab.
+    const shouldResultAppPathBeCached = !isApk || (isApk && isUrl);
+
     if (!isResultAppPathAlreadyCached) {
       if (shouldResultAppPathBeCached) {
-        pathInCache = isApk ? appPath : await this.adb.extractUniversalApk(appPath);
+        // .zip, .aab or downloaded .apk
+        let unzippedAppPath;
+        let isUnzippedAppPath = false;
```

Comment:
do we need this intermediate variable? Cannot we use `unzippedAppPath` for this purpose?

Reply:
not necessary. that was just my preference :)